### PR TITLE
[RFC] vim-patch:7.4.574

### PR DESCRIPTION
**Pull request by buztard**
```
Problem:    No error for eval('$').
Solution:   Check for empty name. (Yasuhiro Matsumoto)
```

https://github.com/vim/vim/commit/v7-4-574

Original patch:

``` diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -7867,7 +7867,7 @@ get_env_tv(arg, rettv, evaluate)
     if (evaluate)
     {
    if (len == 0)
-           return FAIL; /* can't be an environment variable */
+       return FAIL; /* invalid empty name */

    cc = name[len];
    name[len] = NUL;
@@ -10117,14 +10117,18 @@ f_eval(argvars, rettv)
     typval_T   *argvars;
     typval_T   *rettv;
 {
-    char_u *s;
+    char_u *s, *p;

     s = get_tv_string_chk(&argvars[0]);
     if (s != NULL)
    s = skipwhite(s);

+    p = s;
     if (s == NULL || eval1(&s, rettv, TRUE) == FAIL)
     {
+   if (p != NULL && !aborting())
+       EMSG2(_(e_invexpr2), p);
+   need_clr_eos = FALSE;
    rettv->v_type = VAR_NUMBER;
    rettv->vval.v_number = 0;
     }
diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    574,
+/**/
     573,
 /**/
     572,
`
```

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -8139,13 +8139,17 @@ static void f_escape(typval_T *argvars, typval_T *rettv)
  */
 static void f_eval(typval_T *argvars, typval_T *rettv)
 {
-  char_u      *s;
+  char_u      *s, *p;
```
Comment (Pyrohh) [inline src/nvim/eval.c]: the declaration of `p` can be moved down to the initialization

File: src/nvim/eval.c [hunk 2]
```diff
@@ -8139,13 +8139,17 @@ static void f_escape(typval_T *argvars, typval_T *rettv)
...
 
   s = get_tv_string_chk(&argvars[0]);
   if (s != NULL)
     s = skipwhite(s);
 
+  p = s;
   if (s == NULL || eval1(&s, rettv, TRUE) == FAIL) {
+    if (p != NULL && !aborting())
+      EMSG2(_(e_invexpr2), p);
```
Comment (Pyrohh) [inline src/nvim/eval.c]: should be braces around if

**Comment by Pyrohh** [comment]
Besides the comments on style, LGTM.

File: src/nvim/eval.c [hunk 3]
```diff
@@ -6364,7 +6364,7 @@ static int get_env_tv(char_u **arg, typval_T *rettv, int evaluate)
 
   if (evaluate) {
    if (len == 0) {
-      return FAIL;  // Can't be an environment variable.
+      return FAIL;  // invalid empty name
```
Comment (glts) [inline src/nvim/eval.c]: Better match the comment style: `// Invalid empty name.`

**Comment by glts** [comment]
LGTM, the bug is gone :+1: 

Check out the style guide http://neovim.io/develop/style-guide.xml

**Comment by buztard** [comment]
Thanks for the style guide hint, will put it under my pillow :). I updated the patch accorting to your comments and also included some minor modifcations to make clint happy.

File: src/nvim/eval.c [hunk 4]
```diff
@@ -8145,11 +8145,17 @@ static void f_eval(typval_T *argvars, typval_T *rettv)
   if (s != NULL)
     s = skipwhite(s);
 
-  if (s == NULL || eval1(&s, rettv, TRUE) == FAIL) {
+  if (s == NULL || eval1(&s, rettv, true) == FAIL) {
+    char_u *p = s;
```
Comment (glts) [inline src/nvim/eval.c]: I think you overshot the goal slightly :)

`p` must be left outside the block, it should keep pointing to the input string even after `eval1`ing `s`. As it is the patch is no longer correct.

Also, you cannot simply replace `TRUE` with `true` and `FALSE` with `false`, these variables have type `int` (and sometimes hold values other than `0` and `1`). Unless the conversion to `bool` is trivial it is perhaps better to leave it for another pull request.

**Comment by buztard** [comment]
@glts damn, you're so right. Looks like my C skills are a bit rusty, sorry for the inconvenience!

**Comment by glts** [comment]
No problem, LGTM now. Thank you.

**Comment by Pyrohh** [comment]
Both of the failing builds are doing so with this error message:
```
/usr/local/bin/nvim: error while loading shared libraries: libvterm.so.0: cannot open shared object file: No such file or directory
```
Haven't seen that before...

**Comment by fwalch** [comment]
@Pyroh That's because `neovim/deps` was not updated after #2465, hope it works now...

**Comment by Pyrohh** [comment]
Merged, thanks @buztard.

**Pull request closed without merging**